server/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java

```diff
@@ -116,55 +119,71 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request) {
-                    // sometimes the request comes in before the local node processed that cluster state
-                    // in such cases we can load it from disk
-                    metaData = IndexMetaData.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY,
-                        nodeEnv.indexPaths(shardId.getIndex()));
-                }
-                if (metaData == null) {
-                    ElasticsearchException e = new ElasticsearchException("failed to find local IndexMetaData");
-                    e.setShard(request.shardId);
-                    throw e;
-                }
 
-                if (indicesService.getShardOrNull(shardId) == null) {
-                    // we don't have an open shard on the store, validate the files on disk are openable
-                    ShardPath shardPath = null;
-                    try {
-                        IndexSettings indexSettings = new IndexSettings(metaData, settings);
-                        shardPath = ShardPath.loadShardPath(logger, nodeEnv, shardId, indexSettings);
-                        if (shardPath == null) {
-                            throw new IllegalStateException(shardId + " no shard path found");
-                        }
-                        Store.tryOpenIndex(shardPath.resolveIndex(), shardId, nodeEnv::shardLock, logger);
-                    } catch (Exception exception) {
-                        final ShardPath finalShardPath = shardPath;
-                        logger.trace(() -> new ParameterizedMessage(
-                                "{} can't open index for shard [{}] in path [{}]",
-                                shardId,
-                                shardStateMetaData,
-                                (finalShardPath != null) ? finalShardPath.resolveIndex() : ""),
-                            exception);
-                        String allocationId = shardStateMetaData.allocationId != null ?
-                            shardStateMetaData.allocationId.getId() : null;
-                        return new NodeGatewayStartedShards(clusterService.localNode(), allocationId, shardStateMetaData.primary,
-                            exception);
+            final IndexShard indexShard = indicesService.getShardOrNull(shardId);
+            if (indexShard != null) {
+                final ShardStateMetaData shardStateMetaData = indexShard.getShardStateMetaData();
+                final String allocationId = shardStateMetaData.allocationId != null ?
+                    shardStateMetaData.allocationId.getId() : null;
+                logger.trace("{} shard state info found: [{}]", shardId, shardStateMetaData);
+                return new NodeGatewayStartedShards(clusterService.localNode(), allocationId, shardStateMetaData.primary);
+            }
+
+            final ShardStateMetaData shardStateMetaData;
+            try (ShardLock ignored = nodeEnv.shardLock(shardId, TimeUnit.SECONDS.toMillis(5))) {
```

Comment:
Hmm, I just spotted this - there are still two calls to `nodeEnv.shardLock` here. TBH I don't know what we should be doing on failure of this one.